File: leet/0142-linked-list-cycle-II/linked-list-cycle-II.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
    if(head == nil || head.Next == nil) {return nil}
    
    cache := make(map[*ListNode]int)
    
    for count := 0; head != nil; count++ {
        if _, ok := cache[head]; ok{
            return head
        } else{
            cache[head] = count
            head = head.Next
        }
    }
    
    return nil
}

File: leet/0142-linked-list-cycle-II/linked-list-cycle-II.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var detectCycle = function(head) {
    if(!head || !head.next) return null

    let fast = head.next.next
    let slow = head.next
    let met = null

    while (fast && slow){
        if(fast == slow) {
            met = fast
            break
        }
        slow = slow.next
        fast = fast.next ? fast.next.next : null
    }

    if(!met) return null

    let loopLength = 1
    fast = fast.next
    while (fast != met){ 
        loopLength += 1 
        fast = fast.next
    }

    fast = head
    slow = head

    while(loopLength > 0){
        fast = fast.next
        loopLength--
    }

    while(fast != slow){
        fast = fast.next
        slow = slow.next
    }
    return fast
};
